### PR TITLE
Add a dedicated configuration file to generate 'src/app/encoder.cpp' …

### PR DESCRIPTION
**Pull request by vivien-apple**
…and 'src/app/chip-zcl-zpro-codec-api.h'

#### Problem

So far `src/app/chip-zcl-zpro-codec-api.h` and `src/app/encoder.cpp` has been generated
using times to times by manually adding the related templates to to the `chip-templates.json` and
removing it before checking-in the code since this code does not really belongs to application folders.

It is error prone. An in fact I broken it with #4155 since attributes appears multiple times because of
the second endpoint.

This PR splits the `zap-templates` folder in two, with one folder for the examples and one fodler for
chip related code.

#### Summary of changes
 * Split `zap-templates` folder in two
 * Split `helper-chip.js` into multiple helpers and move those helpers where they belongs
 * Add a new `scripts/tools/zap_generate_chip.sh` script to generate the `src/app` content
 * Add `scripts/tools/zap_generate_chip.sh` to `scripts/tools/zap_regen_all.py`
 * Create a partial for the licensing `header` instead of relying on js for it
 * Fix the attribute issue mentioned above
 * Remove `asChipUnderlyingType` since it relies on `asUnderlyingZclType` and we should be able to fix the issue directly once #4192 lands and it really does not introduce any problem to remove it

### CODE REVIEW

**Comment by vivien-apple** [comment]
Also the out-of-sync issue will only be worse when #4126 lands. This PR creates a place to host the `CHIPClusters.zapt` and `CHIPClusters-src.zapt` too.

File: examples/all-clusters-app/all-clusters-common/gen/af-structs.h [hunk 1]
```diff
@@ -88,7 +87,7 @@ typedef struct _DebtPayload
 typedef struct _DeviceInformationRecord
 {
     uint64_t ieeeAddress;
-    chip::EndpointId endpointId;
+    uint8_t endpointId;
```
Comment (bzbarsky-apple) [inline examples/all-clusters-app/all-clusters-common/gen/af-structs.h:90]: This change, and similar, does not seem right.

File: src/app/zap-templates/templates/app/af-structs.zapt [hunk 2]
```diff
@@ -13,7 +13,7 @@
 // Struct for {{label}}
 typedef struct _{{asType label}} {
 {{#zcl_struct_items}}
-{{ident}}{{asChipUnderlyingType label type}} {{asSymbol label}};
+{{ident}}{{asUnderlyingZclType type}} {{asSymbol label}};
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/templates/app/af-structs.zapt:16]: Is this why the generated types change from `chip::....` to raw integers?

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/app/zap-templates/templates/app/af-structs.zapt (hunk 2 above)
Comment (vivien-apple) [reply src/app/zap-templates/templates/app/af-structs.zapt]: Yes. That was it. I have reverted this change.

**Review by bzbarsky-apple — APPROVED** [review]
Thank you!

**Review by andy31415 — APPROVED** [review]

**Review by rwalker-apple — APPROVED** [review]

**Comment by rwalker-apple** [comment]
@hawk248 @BroderickCarlin @jelderton @saurabhst ?

**Review by saurabhst — APPROVED** [review]

**Pull request merged**